zipline/pipeline/factors/technical.py

```diff
@@ -683,3 +679,115 @@ def compute(self, today, assets, out, highs, lows, closes):
+
+
+class MovingAverageConvergenceDivergence(_ExponentialWeightedFactor):
+    """
+    Moving Average Convergence/Divergence (MACD)
+    https://en.wikipedia.org/wiki/MACD
+
+    A technical indicator originally developed by Gerald Appel in the late
+    1970's. MACD shows the relationship between two moving averages and
+    reveals changes in the strength, direction, momentum, and duration of a
+    trend in a stock's price.
+
+    **Default Inputs:** :data:`zipline.pipeline.data.USEquityPricing.close`
+    **Default Window Length:** Window length is automatically calculated as the
+    sum of slow_period and signal_period.
+
+    Parameters
+    ----------
+    fast_period : int >= 0, <= window_length
+        The window length for the "fast" EWMA. Default is 12.
+    slow_period : int >= 0, <= window_length
+        The window length for the "slow" EWMA. Default is 26.
+    signal_period' : int >= 0, <= slow_period
+        The window length for the signal line. Default is 9.
+
+    Returns
+    -------
+    MACD: The difference between "fast" EWMA and "slow" EWMA.
+    signal: The EWMA of the MACD line using `signal_period` as span.
+    hist: Difference between MACD and signal. (Divergence series)
+    """
+    inputs = [USEquityPricing.close]
+    params = ('fast_period', 'slow_period', 'signal_period')
+    outputs = ('MACD', 'signal', 'hist')
+
+    def __new__(cls,
+                fast_period=12,
+                slow_period=26,
+                signal_period=9,
+                *args,
+                **kwargs):
+        return super(MovingAverageConvergenceDivergence, cls).__new__(
+            cls,
+            fast_period=fast_period,
+            slow_period=slow_period,
+            signal_period=signal_period,
+            window_length=slow_period + signal_period - 1,
+            *args, **kwargs
+        )
+
+    def calculate_ewma(self, data, length):
+        decay_rate = 1.0 - (2.0 / (1.0 + length))
+        return average(data,
+                       axis=1,
+                       weights=self.weights(length, decay_rate))
+
+    def calculate_macd(self, col):
+        slow_EWMA = self.calculate_ewma(
+            rolling_window(
+                col,
+                self.params['slow_period']
+            ),
+            self.params['slow_period'])
+        fast_EWMA = self.calculate_ewma(
+            rolling_window(
+                col,
+                self.params['fast_period']
+            )[-self.params['signal_period']:],
+            self.params['fast_period'])
+        macd = fast_EWMA - slow_EWMA
+        signal_line = self.calculate_ewma(
+            macd.reshape(-1, self.params['signal_period']),
+            self.params['signal_period'])
+        hist = macd[-1] - signal_line
+        return macd[-1], signal_line[-1], hist[-1]
+
+    def compute(self, today, assets, out, close, fast_period, slow_period,
+                signal_period):
+        macd, sig, hist = zip(*map(self.calculate_macd, close.T))
```

Comment:
Why are calling `calclulate_macd` separately on each column of `close`?  It seems like we should be able to do it vectorized on the whole 2d array?